File: actionbar/src/main/java/mobi/cangol/mobile/actionbar/view/ActionMenuView.java
package mobi.cangol.mobile.actionbar.view;

import android.content.Context;
import android.graphics.drawable.BitmapDrawable;
import android.os.Build;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.View.OnLongClickListener;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.PopupWindow;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;

import mobi.cangol.mobile.actionbar.ActionMenu;
import mobi.cangol.mobile.actionbar.ActionMenuItem;
import mobi.cangol.mobile.actionbar.R;

/**
 * @author Cangol
 */
public class ActionMenuView extends LinearLayout implements OnClickListener, OnLongClickListener {

    private LayoutInflater mInflater;
    private LinearLayout mActionsView;
    private LinearLayout mPopupActionsView;
    private ImageView mMoreButton;
    private PopupWindow mPopuMenu;
    private ActionMenu mActionMenu;
    private OnActionClickListener mOnActionClickListener;
    private int mShowActions = 0;

    public ActionMenuView(Context context) {
        super(context);
        initViews(context);
    }

    public ActionMenuView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initViews(context);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public ActionMenuView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initViews(context);
    }

    public ActionMenuView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initViews(context);
    }

    public void setActionMenu(ActionMenu mActionMenu) {
        this.mActionMenu = mActionMenu;
    }

    private void initViews(Context context) {
        mInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        mInflater.inflate(R.layout.actionbar_menu_view, this, true);
        mActionsView = (LinearLayout) this.findViewById(R.id.actionbar_menu_actions);
        mMoreButton = (ImageView) this.findViewById(R.id.actionbar_menu_more);
        initPopupMenu(context);
    }

    private void initPopupMenu(Context context) {
        View popuLayout = mInflater.inflate(R.layout.actionbar_popup_actions, null);
        mPopupActionsView = (LinearLayout) popuLayout.findViewById(R.id.actionbar_popup_actions);
        int width = (int) (180 * context.getResources().getDisplayMetrics().density);
        mPopuMenu = new PopupWindow(popuLayout, width, LayoutParams.WRAP_CONTENT, true);
        mPopuMenu.setBackgroundDrawable(new BitmapDrawable());
        mMoreButton.setOnClickListener(new OnClickListener() {

            @Override
            public void onClick(View v) {
                mPopuMenu.showAsDropDown(mMoreButton);
            }

        });
    }

    @Override
    public void onClick(View view) {
        final Object tag = view.getTag();
        if (tag instanceof ActionMenuItem) {
            final ActionMenuItem action = (ActionMenuItem) tag;
            if (mPopuMenu != null && mPopuMenu.isShowing()) {
                mPopuMenu.dismiss();
            }
            if (mOnActionClickListener != null) {
                mOnActionClickListener.onActionClick(action);
            }
        }
    }

    @Override
    public boolean onLongClick(View view) {
        final Object tag = view.getTag();
        if (tag instanceof ActionMenuItem) {
            final ActionMenuItem action = (ActionMenuItem) tag;

            Toast toast = null;
            if (action.getText() != -1)
                toast = Toast.makeText(this.getContext(), action.getText(), Toast.LENGTH_SHORT);

            int left = ((mActionMenu.size() - mShowActions) > 0 ? mMoreButton.getWidth() : 0) + view.getWidth() * (mShowActions) - view.getLeft();
            toast.setGravity(Gravity.RIGHT | Gravity.TOP, left, this.getBottom());
            toast.show();
        }
        return true;
    }

    public void addActions(List<ActionMenuItem> actions) {
        for (int i = 0; i < actions.size(); i++) {
            addAction(actions.get(i));
        }
    }

    /**
     * Adds a new {@link ActionMenuItem}.
     *
     * @param action the action to add
     */
    public void addAction(ActionMenuItem action) {

        //final int size = mActionMenu.size();
        //int showActions=size-c;
        if (action.isShow()) {
            if (mShowActions < 1) {
                mActionsView.addView(action.isIcon() ? inflateActionIcon(action) : inflateActionText(action));
                mMoreButton.setVisibility(View.GONE);
                mShowActions++;
            } else if (mShowActions < 2 && mActionMenu.size() == 2) {
                mActionsView.addView(action.isIcon() ? inflateActionIcon(action) : inflateActionText(action));
                mMoreButton.setVisibility(View.GONE);
                mShowActions++;
            } else {
                if (mShowActions == 2) {
                    mActionsView.removeViewAt(1);
                    mPopupActionsView.addView(inflateMenuAction(mActionMenu.getAction(2)));
                    mMoreButton.setVisibility(View.VISIBLE);
                }
                mPopupActionsView.addView(inflateMenuAction(action));
                mMoreButton.setVisibility(View.VISIBLE);
            }
        } else {
            if (mShowActions == 2) {
                mActionsView.removeViewAt(1);
                mPopupActionsView.addView(inflateMenuAction(mActionMenu.getAction(2)));
                mMoreButton.setVisibility(View.VISIBLE);
            }
            mPopupActionsView.addView(inflateMenuAction(action));
            mMoreButton.setVisibility(View.VISIBLE);

        }
    }

    /**
     * Removes all action views from this action bar
     */
    public void removeAllActions() {
        mShowActions = 0;
        mActionsView.removeAllViews();
        mPopupActionsView.removeAllViews();
        mMoreButton.setVisibility(View.GONE);
    }


    private View inflateActionIcon(ActionMenuItem action) {
        View view = mInflater.inflate(R.layout.actionbar_item_icon, mActionsView, false);

        ImageView labelView =
                (ImageView) view.findViewById(R.id.actionbar_item);
        labelView.setImageResource(action.getDrawable());
        view.setId(action.getId());
        view.setTag(action);
        view.setOnClickListener(this);
        view.setOnLongClickListener(this);
        return view;
    }

    private View inflateActionText(ActionMenuItem action) {
        View view = mInflater.inflate(R.layout.actionbar_item_text, mActionsView, false);

        TextView labelView = (TextView) view.findViewById(R.id.actionbar_item);
        labelView.setText(action.getText());

        view.setId(action.getId());
        view.setTag(action);
        view.setOnClickListener(this);
        view.setOnLongClickListener(this);
        return view;
    }

    private View inflateMenuAction(ActionMenuItem action) {
        View view = mInflater.inflate(R.layout.actionbar_popup_item, mPopupActionsView, false);
        TextView labelView = (TextView) view.findViewById(R.id.actionbar_popup_item_text);
        /**
         if(action.getDrawable()>0){
         Drawable img=getResources().getDrawable(action.getDrawable());
         img.setBounds(0, 0, img.getIntrinsicWidth(), img.getIntrinsicHeight());
         labelView.setCompoundDrawables(img, null, null, null);
         }**/
        labelView.setText(action.getText());

        view.setId(action.getId());
        view.setTag(action);
        view.setOnClickListener(this);
        view.setOnLongClickListener(this);
        return view;
    }


    public void setOnActionClickListener(OnActionClickListener mOnActionClickListener) {
        this.mOnActionClickListener = mOnActionClickListener;
    }

    public interface OnActionClickListener {

        boolean onActionClick(ActionMenuItem action);
    }
}

File: actionbar/src/main/java/mobi/cangol/mobile/actionbar/view/ActionMenuView.kt
package mobi.cangol.mobile.actionbar.view

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.support.annotation.RequiresApi
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.View.OnLongClickListener
import android.widget.*
import mobi.cangol.mobile.actionbar.ActionMenu
import mobi.cangol.mobile.actionbar.ActionMenuItem
import mobi.cangol.mobile.actionbar.R

/**
 * @author Cangol
 */
class ActionMenuView : LinearLayout, OnClickListener, OnLongClickListener {

    private var mInflater: LayoutInflater? = null
    private var mActionsView: LinearLayout? = null
    private var mPopupActionsView: LinearLayout? = null
    private var mMoreButton: ImageView? = null
    private var mPopuMenu: PopupWindow? = null
    private var mActionMenu: ActionMenu? = null
    private var mOnActionClickListener: OnActionClickListener? = null
    private var mShowActions = 0

    constructor(context: Context) : super(context) {
        initViews(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initViews(context)
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        initViews(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initViews(context)
    }

    fun setActionMenu(mActionMenu: ActionMenu) {
        this.mActionMenu = mActionMenu
    }

    private fun initViews(context: Context) {
        mInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        mInflater?.inflate(R.layout.actionbar_menu_view, this, true)
        mActionsView = this.findViewById(R.id.actionbar_menu_actions)
        mMoreButton = this.findViewById(R.id.actionbar_menu_more)
        initPopupMenu(context)
    }

    private fun initPopupMenu(context: Context) {
        val popuLayout = mInflater?.inflate(R.layout.actionbar_popup_actions, null)
        mPopupActionsView = popuLayout?.findViewById(R.id.actionbar_popup_actions)
        val width = (180 * context.resources.displayMetrics.density).toInt()
        mPopuMenu = PopupWindow(popuLayout, width, LayoutParams.WRAP_CONTENT, true)
        mPopuMenu?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        mMoreButton?.setOnClickListener { mPopuMenu?.showAsDropDown(mMoreButton) }
    }

    override fun onClick(view: View) {
        val tag = view.tag
        if (tag is ActionMenuItem) {
            if (mPopuMenu?.isShowing==true) {
                mPopuMenu?.dismiss()
            }
            mOnActionClickListener?.onActionClick(tag)
        }
    }

    override fun onLongClick(view: View): Boolean {
        val tag = view.tag
        if (tag is ActionMenuItem) {

            var toast: Toast? = null
            if (tag.text != -1)
                toast = Toast.makeText(this.context.applicationContext, tag.text, Toast.LENGTH_SHORT)

            val left = (if (mActionMenu!!.size() - mShowActions > 0) mMoreButton!!.width else 0) + view.width * mShowActions - view.left

            if (toast != null) {
                toast.setGravity(Gravity.RIGHT or Gravity.TOP, left, this.bottom)
                toast.show()
            }
        }
        return true
    }

    fun addActions(actions: List<ActionMenuItem>) {
        for (menuItem in actions) {
            addAction(menuItem)
        }
    }

    /**
     * Adds a new [ActionMenuItem].
     *
     * @param action the action to add
     */
    fun addAction(action: ActionMenuItem) {
        if (action.isShow) {
            if (mShowActions < 1) {
                mActionsView?.addView(if (action.isIcon) inflateActionIcon(action) else inflateActionText(action))
                mMoreButton?.visibility = View.GONE
                mShowActions++
            } else if (mShowActions < 2 && mActionMenu?.size() == 2) {
                mActionsView?.addView(if (action.isIcon) inflateActionIcon(action) else inflateActionText(action))
                mMoreButton?.visibility = View.GONE
                mShowActions++
            } else {
                if (mShowActions == 2) {
                    mActionsView?.removeViewAt(1)
                    mPopupActionsView?.addView(mActionMenu?.getAction(2)?.let { inflateMenuAction(it) })
                    mMoreButton?.visibility = View.VISIBLE
                }
                mPopupActionsView?.addView(inflateMenuAction(action))
                mMoreButton?.visibility = View.VISIBLE
            }
        } else {
            if (mShowActions == 2) {
                mActionsView?.removeViewAt(1)
                mPopupActionsView?.addView(mActionMenu?.getAction(2)?.let { inflateMenuAction(it) })
                mMoreButton?.visibility = View.VISIBLE
            }
            mPopupActionsView?.addView(inflateMenuAction(action))
            mMoreButton?.visibility = View.VISIBLE

        }
    }

    /**
     * Removes all action views from this action bar
     */
    fun removeAllActions() {
        mShowActions = 0
        mActionsView?.removeAllViews()
        mPopupActionsView?.removeAllViews()
        mMoreButton?.visibility = View.GONE
    }


    private fun inflateActionIcon(action: ActionMenuItem): View {
        val view = mInflater?.inflate(R.layout.actionbar_item_icon, mActionsView, false)

        val labelView = view?.findViewById<View>(R.id.actionbar_item) as ImageView
        labelView.setImageResource(action.drawable)
        view.id = action.id
        view.tag = action
        view.setOnClickListener(this)
        view.setOnLongClickListener(this)
        return view
    }

    private fun inflateActionText(action: ActionMenuItem): View {
        val view = mInflater?.inflate(R.layout.actionbar_item_text, mActionsView, false)

        val labelView = view?.findViewById(R.id.actionbar_item) as TextView
        labelView.setText(action.text)

        view.id = action.id
        view.tag = action
        view.setOnClickListener(this)
        view.setOnLongClickListener(this)
        return view
    }

    private fun inflateMenuAction(action: ActionMenuItem): View {
        val view = mInflater?.inflate(R.layout.actionbar_popup_item, mPopupActionsView, false)
        val labelView = view?.findViewById(R.id.actionbar_popup_item_text) as TextView
        /**
         * if(action.getDrawable()>0){
         * Drawable img=getResources().getDrawable(action.getDrawable());
         * img.setBounds(0, 0, img.getIntrinsicWidth(), img.getIntrinsicHeight());
         * labelView.setCompoundDrawables(img, null, null, null);
         * } */
        labelView.setText(action.text)

        view.id = action.id
        view.tag = action
        view.setOnClickListener(this)
        view.setOnLongClickListener(this)
        return view
    }


    fun setOnActionClickListener(mOnActionClickListener: OnActionClickListener) {
        this.mOnActionClickListener = mOnActionClickListener
    }

    interface OnActionClickListener {

        fun onActionClick(action: ActionMenuItem): Boolean
    }
}
